client/diff.go

```diff
@@ -0,0 +1,273 @@
+	RolesRemoved   []data.Role
+	RolesUpdated   []data.Role
+}
+
+// NewDiff returns the different between two versions of the same TUF repo
+func NewDiff(gun data.GUN, baseURL string, rt http.RoundTripper, hash1, hash2 string) (Diff, error) {
+	remoteStore, err := getRemoteStore(baseURL, gun, rt)
+	if err != nil {
+		// baseURL is syntactically invalid
+		return Diff{}, err
+	}
+
+	first, err := setupClient(gun, hash1, remoteStore)
+	if err != nil {
+		return Diff{}, err
+	}
+	second, err := setupClient(gun, hash2, remoteStore)
+	if err != nil {
+		return Diff{}, err
+	}
+
+	// download the relevant versions of the repos now that we have the bootstrapping
+	// setup.
+	// N.B. lower case update will fail out immediately rather than attempting
+	//      to download a newer root
+	logrus.Debug("starting updates")
+	err = first.update()
+	if err != nil {
+		return Diff{}, errors.New("failed to download repo version " + hash1)
+	}
+	logrus.Debug("succeeded updating first")
+	err = second.update()
+	if err != nil {
+		return Diff{}, errors.New("failed to download repo version " + hash2)
+	}
+	logrus.Debug("succeeded updated second")
+
+	repoFirst, _, err := first.newBuilder.Finish()
+	repoSecond, _, err := second.newBuilder.Finish()
+
+	return diff(repoFirst, repoSecond)
+}
+
+func setupClient(gun data.GUN, tsHash string, remote store.RemoteStore) (*tufClient, error) {
+	hashBytes, err := hex.DecodeString(tsHash)
+	if err != nil {
+		return nil, err
+	}
+
+	root, err := findRoot(hashBytes, remote)
+
+	oldBuilder := tuf.NewRepoBuilder(gun, nil, trustpinning.TrustPinConfig{})
+	if err := oldBuilder.Load(data.CanonicalRootRole, root, 0, true); err != nil {
+		return nil, err
+	}
+	newBuilder := tuf.NewRepoBuilder(gun, nil, trustpinning.TrustPinConfig{})
+	if err := newBuilder.Load(data.CanonicalRootRole, root, 0, true); err != nil {
+		return nil, err
+	}
+
+	tClient := newTufClient(oldBuilder, newBuilder, remote, store.NewMemoryStore(nil))
+	tClient.tsChecksum = hashBytes
+	return tClient, nil
+}
+
+func diff(first, second *tuf.Repo) (Diff, error) {
+	res := Diff{}
+	err := diffTargets(&res, first, second)
+	if err != nil {
+		return res, err
+	}
+	err = diffRoles(&res, first, second)
+	return res, err
+}
+
+func diffRoles(res *Diff, first, second *tuf.Repo) error {
+	firstRoles := first.GetAllLoadedRoles()
+	lookupTable := make(map[string]*data.Role)
+	for _, role := range firstRoles {
+		lookupTable[role.Name.String()] = role
+	}
+	secondRoles := second.GetAllLoadedRoles()
+	for _, role := range secondRoles {
+		found, ok := lookupTable[role.Name.String()]
+		if !ok {
+			res.RolesAdded = append(res.RolesAdded, *role)
+			continue
+		}
+		delete(lookupTable, role.Name.String())
+		if !equivalentRoles(role, found) {
+			res.RolesUpdated = append(res.RolesUpdated, *role)
+		}
+	}
+	for _, role := range lookupTable {
+		res.RolesRemoved = append(res.RolesRemoved, *role)
+	}
+	return nil
+}
+
+func equivalentRoles(first, second *data.Role) bool {
```

Comment:
Could we add a comment that we check for paths, thresholds, key ids, and _not_ versions because that isn't a meaningful update in the role?